services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java

```diff
@@ -0,0 +1,261 @@
+        } else {
+            // in ID-less & Password-less Scenario
+            // NOP
+        }
+        form.setAttribute(WebAuthnConstants.IS_USER_IDENTIFIED, Boolean.toString(isUserIdentified));
+
+        // read options from authenticator config
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+        String userVerificationRequirement = config.getConfig().get(WebAuthnAuthenticatorFactory.USER_VERIFICATION_REQUIREMENT);
+        if (WebAuthnConstants.OPTION_NOT_SPECIFIED.equals(userVerificationRequirement)) userVerificationRequirement = Constants.DEFAULT_WEBAUTHN_POLICY_NOT_SPECIFIED;
+        form.setAttribute(WebAuthnConstants.USER_VERIFICATION, userVerificationRequirement);
+
+        context.challenge(form.createForm("webauthn-authenticate.ftl"));
+    }
+
+    public void action(AuthenticationFlowContext context) {
+        MultivaluedMap<String, String> params = context.getHttpRequest().getDecodedFormParameters();
+
+        // receive error from navigator.credentials.get()
+        String errorMsgFromWebAuthnApi = params.getFirst(WebAuthnConstants.ERROR);
+        if (errorMsgFromWebAuthnApi != null && !errorMsgFromWebAuthnApi.isEmpty()) {
+            setErrorResponse(context, ERR_WEBAUTHN_API_GET, errorMsgFromWebAuthnApi);
+            return;
+        }
+
+        String baseUrl = UriUtils.getOrigin(context.getUriInfo().getBaseUri());
+        String rpId = context.getUriInfo().getBaseUri().getHost();
+
+        Origin origin = new Origin(baseUrl);
+        Challenge challenge = new DefaultChallenge(context.getAuthenticationSession().getAuthNote(WebAuthnConstants.AUTH_CHALLENGE_NOTE));
+        ServerProperty server = new ServerProperty(origin, rpId, challenge, null);
+
+        byte[] credentialId = Base64Url.decode(params.getFirst(WebAuthnConstants.CREDENTIAL_ID));
+        byte[] clientDataJSON = Base64Url.decode(params.getFirst(WebAuthnConstants.CLIENT_DATA_JSON));
+        byte[] authenticatorData = Base64Url.decode(params.getFirst(WebAuthnConstants.AUTHENTICATOR_DATA));
+        byte[] signature = Base64Url.decode(params.getFirst(WebAuthnConstants.SIGNATURE));
+
+        String userId = params.getFirst(WebAuthnConstants.USER_HANDLE);
+        boolean isUVFlagChecked = false;
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+        String userVerificationRequirement = config.getConfig().get(WebAuthnAuthenticatorFactory.USER_VERIFICATION_REQUIREMENT);
+        if (WebAuthnConstants.OPTION_REQUIRED.equals(userVerificationRequirement)) isUVFlagChecked = true;
+
+        // existing User Handle means that the authenticator used Resident Key supported public key credential
+        if (userId == null || userId.isEmpty()) {
+            // Resident Key not supported public key credential was used
+            // so rely on the user that has already been authenticated
+            userId = context.getUser().getId();
```

Comment:
I think NPE does not happen in the following block :

```java
        // existing User Handle means that the authenticator used Resident Key supported public key credential
        if (userId == null || userId.isEmpty()) {
            // Resident Key not supported public key credential was used
            // so rely on the user that has already been authenticated
            userId = context.getUser().getId();
        } else {
```

The reason is as follows:

When NPE occurs, the follwing two condition need be met :

1. `context.getUser()` is null
   which means that no user is not yet authenticated on keycloak

2. `userId` is null or empty
   which means that `navigator.credentials.get()` returns successful response, but it does not include `User Handle`.

Considering 1, in `WebAutnAuthenticator.authenticate()`, keycloak can not set the `navigator.credentials.get()`'s parameter `PublicKeyCredentialRequestOptions.allowCredentials`.

`Credential ID` has to be set to this, however, keycloak can not determine whose `Credential ID` has to set to it because keycloak does not yet authenticate any user at that time.

Therefore, the browser call `navigator.credentials.get()` without the parameter `PublicKeyCredentialRequestOptions.allowCredentials`.

After calling `navigator.credentials.get()`, in order for keycloak to reach the codes above, the following conditions has to be met :

- `navigator.credentials.get()` returns a successful response
- `userId` is null or empty

To get a successful response by calling `navigator.credentials.get()` without `PublicKeyCredentialRequestOptions.allowCredentials`, the following condition need to be met at least :

- The webauthn authenticator has the capability of `Resident Key`
- The webauthn authenticator has already created the public key credential as `client-side-resident public key credential` on `navigator.credentials.create()`

and, such the webauthn authenticator returns successful response with `User Handle`. (refer to https://www.w3.org/TR/webauthn/#op-make-cred 7-4-4)

Considering all above, if `context.getUser()` is null, `userId` is not null on the codes above.

Reply:
Makes sense, Thanks!